src/main/java/org/owasp/esapi/Encoder.java

```diff
@@ -107,16 +109,42 @@
  * in other applications that could access that data in the database.
  * </li>
  * <li><b>Avoiding multiple <i>nested</i> contexts:</b>
- * A really tricky situation to get correct is hen there are multiple nested
+ * A really tricky situation to get correct is when there are multiple nested
  * encoding contexts. But far, the most common place this seems to come up is
- * untrusted URLs used in JavaScript. How should you handle that? Well, to be
- * honest, the best way is to rewrite your code to avoid it.  An example of
+ * untrusted URLs used in JavaScript. How should you handle that? Well,
+ * the best way is to rewrite your code to avoid it!  An example of
  * this that is well worth reading may be found at
  * <a href="https://lists.owasp.org/pipermail/esapi-dev/2012-March/002090"
  * target="_blank" rel="noopener noreferrer">ESAPI-DEV mailing list archives:
  * URL encoding within JavaScript</a>. Be sure to read the entire thread.
  * The question itself is too nuanced to be answered in Javadoc, but now,
- * hopefully you are at least aware of the potential pitfalls.
+ * hopefully you are at least aware of the potential pitfalls. There is little
+ * avaiable research or examples on how to do output encodeing when multiple
+ * mixed encodings are required, although one that you may find useful is
+ * <a href="https://arxiv.org/pdf/1804.01862.pdf" target="_blank"
+ * rel="noopener noreferrer">
+ * Automated Detecting and Repair of Cross-SiteScripting Vulnerabilities through Unit Testing</a>
+ * It at least discusses a few of the common errors involved in multiple mixed
+ * encoding contexts.
+ * </li><li><b>A word about unit testing:</b>
+ * Unit testing this is hard. You may be satisfied with stopped after you have
+ * tested against the ubiquitous XSS test case of
+ * <pre>
+ *      &lt;/script&gt;alert(1)&lt;/script&gt;
+ * </pre>
+ * or similiar simplistic XSS attack payloads and if that is properly encoded
+ * (or, you don't see an alert box popped in your browser), you consider it
+ * "problem fixed", consider the unit testing sufficient. Unfortunately, that
+ * minimalist testing may not always detect places were you used the wrong decoder. You need to do better.
+ * Fortunately, the aforementioned link,
+ * <a href="https://arxiv.org/pdf/1804.01862.pdf" target="_blank"
+ * rel="noopener noreferrer">
+ * Automated Detecting and Repair of Cross-SiteScripting Vulnerabilities through Unit Testing</a>
```

Comment:
@jeremiahjstacey should we resurrect our old idea of a security-based fuzzing JUnit framework?  This part of the comments reminds me of some of our old talks about that.  (For @kwwall 's sake it amounts to taking a rather large collection of XSS attacks and automating attacking parameters with them.)

Reply:
I often had thought it would be useful to take many of our JUnit tests and make them data-driven so we could just put the test data and the expected outcomes into a file. If we did that, it would make fuzzing easier. I have wanted to do this for years for the Encryptor tests so I could take a more standard cryptographic test vectors approach. It makes it really easy to add new test cases. To my knowledge, the only place we are using an approach like that is with all those tests that you (Matt) added to testing the URL validation / canonicalization tests.

Reply:
Yep we're tracking together.  In the past we built a custom JUnit test that started with an input file that collected *all* the attack inputs I could find, and then would include an expected value as to whether or not that attack should pass or fail whatever filter we were working on.  At the time, we had talked about creating a more generalized framework that folks could drop into their own codebase.  The issue here is that it requires a great deal of careful judgment and to be blunt, you can't trust THAT to just any programmer off the street.